test/integration/controlplane/transformation/secrets_transformation_test.go

```diff
@@ -99,6 +120,452 @@ func TestSecretsShouldBeTransformed(t *testing.T) {
+	secrets := []string{"corrupt-a", "corrupt-b", "corrupt-c"}
+
+	tests := []struct {
+		runner listCorruptObjectTestRunner
+	}{
+		{
+			runner: listCorruptObjectTestRunner{
+				secrets:        secrets,
+				featureEnabled: true,
+				encryptionBrokenFn: func(t *testing.T, got apierrors.APIStatus) bool {
+					return got.Status().Reason == metav1.StatusReasonInternalError &&
+						strings.Contains(got.Status().Message, "data from the storage is not transformable")
+				},
+				listAfter: wantAPIStatusError{
+					reason:          metav1.StatusReasonStoreReadError,
+					messageContains: "failed to read one or more secrets from the storage",
+					more: func(t *testing.T, err apierrors.APIStatus) {
+						t.Helper()
+
+						details := err.Status().Details
+						if details == nil {
+							t.Errorf("expected Details in APIStatus, but got: %#v", err)
+							return
+						}
+						if want, got := len(secrets), len(details.Causes); want != got {
+							t.Errorf("expected to have %d in APIStatus, but got: %d", want, got)
+						}
+						for _, cause := range details.Causes {
+							if want, got := metav1.CauseTypeUnexpectedServerResponse, cause.Type; want != got {
+								t.Errorf("expected to cause type to be %s, but got: %s", want, got)
+							}
+						}
+						for _, want := range secrets {
+							var found bool
+							for _, got := range details.Causes {
+								if strings.HasSuffix(got.Field, want) {
+									found = true
+									break
+								}
+							}
+							if !found {
+								t.Errorf("want key: %q in the Fields: %#v", want, details.Causes)
+							}
+						}
+					},
+				},
+			},
+		},
+		{
+			runner: listCorruptObjectTestRunner{
+				secrets:        secrets,
+				featureEnabled: false,
+				encryptionBrokenFn: func(t *testing.T, got apierrors.APIStatus) bool {
+					return got.Status().Reason == metav1.StatusReasonInternalError &&
+						strings.Contains(got.Status().Message, "cipher: message authentication failed")
+				},
+				listAfter: wantAPIStatusError{
+					reason:          metav1.StatusReasonInternalError,
+					messageContains: "unable to transform key",
+				},
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("%s/%t", string(genericfeatures.AllowUnsafeMalformedObjectDeletion), tc.runner.featureEnabled), func(t *testing.T) {
+			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.AllowUnsafeMalformedObjectDeletion, tc.runner.featureEnabled)
+			tc.runner.Run(t)
+		})
+	}
+}
+
+type listCorruptObjectTestRunner struct {
+	featureEnabled bool
+	// secrets that are created before encryption breaks
+	secrets []string
+	// whether encryption broke after the config change
+	encryptionBrokenFn func(t *testing.T, got apierrors.APIStatus) bool
+	// what we expect for LIST on the corrupt objects after encryption has broken
+	listAfter verifier
+}
+
+func (testrunner listCorruptObjectTestRunner) Run(t *testing.T) {
```

Comment:
Same comment regarding the struct and the Run method.

Reply:
yes, done